### PR TITLE
Store list of files for the sample into the database

### PR DESCRIPTION
**Pull request by blinkseb**
With this PR, the file of files is stored into the database in addition to the path:
- We use the list of files given by crab. It includes all the jobs done
- From this list of files, the number of events and the sum of event weight are computed

If the job is not done, a warning is issued by nothing else is done. The sample can be updated any time after by simply re-executing the script. Since the list of files is stored in the database, it's fine if some files are added to the output directory after the push to the database.

Also, I've made the code independent of ingrid. It can now be ran anywhere, at the condition you have access to the database.

PS: This does not work for the moment as the database on ingrid is not updated. This is for the moment only ready for a review. The needed changes to SAMADhi can be found here: https://github.com/blinkseb/SAMADhi/commit/4330a33c5b245ee4d08038f2ac7e4879c0408721

### CODE REVIEW

File: scripts/runPostCrab.py [hunk 1]
```diff
@@ -152,18 +189,19 @@ def main():
...
     # Since the API outputs AND prints the same data, hide whatever is printed on screen
     saved_stdout, saved_stderr = sys.stdout, sys.stderr
     if not options.debug:
         sys.stdout = sys.stderr = open(os.devnull, "w")
     taskdir = os.path.join(workArea, 'crab_' + requestName)
-    # list logs
-    log_files = crabCommand('getlog', '--dump', dir = taskdir )
     # list output
     output_files = crabCommand('getoutput', '--dump', dir = taskdir )
+    output_files_xrootd = crabCommand('getoutput', '--dump', '--xrootd', dir = taskdir )
```
Comment (OlivierBondu) [inline scripts/runPostCrab.py]: why the `--xrootd` ?

Comment (blinkseb) [inline scripts/runPostCrab.py]: Mainly to be site independent. You don't have to know where the file is stored anymore. I first tried with the default `pfn`, but root does not handle `srm://` url.

Comment (blinkseb) [inline scripts/runPostCrab.py]: And both `lfn` and `pfn` are stored in the database

Comment (blinkseb) [inline scripts/runPostCrab.py]: Not sure if both are needed

File: scripts/runPostCrab.py [hunk 2]
```diff
@@ -174,6 +212,39 @@ def main():
...
+    dataset_nselected = 0
+    for f in files:
+        (sumw, entries) = get_file_data(f['pfn'])
+        if sumw == 0:
+            print("Warning: failed to retrieve sum of event weight for %r" % f['lfn'])
+
+        dataset_sumw += sumw
+
+        if entries == 0:
+            print("Warning: failed to retrieve number of entries for %r" % f['lfn'])
```
Comment (OlivierBondu) [inline scripts/runPostCrab.py:249]: mmm, QCD is here now :worried:

Comment (blinkseb) [inline scripts/runPostCrab.py:249]: What is your concern here?

Comment (blinkseb) [inline scripts/runPostCrab.py:249]: Oh I see

Comment (blinkseb) [inline scripts/runPostCrab.py:249]: So let change the DB field to nullable, and store NULL if it fails to get the number of entries instead of 0

Comment (OlivierBondu) [inline scripts/runPostCrab.py:249]: sounds good

**Comment by blinkseb** [comment]
I've setup a test database on ingrid. It would nice if someone can test the workflow and see if the files correctly end-up in the database.

You'll need to install SAMADhi from my fork as well as this PR:

```
git remote add blinkseb https://github.com/blinkseb/SAMADhi.git
git fetch blinkseb
git checkout files
```

Then, just edit the `scripts/SAMADhi.py` files with the proper credentials, and change the database from `llbb` to `llbb_test`. Do not forget to rebuild everything.

**Comment by OlivierBondu** [comment]
testing

**Comment by OlivierBondu** [comment]
so, I tested on an incomplete and partially failed QCD task (seemed perfect for testing):

```
>crab status tasks/crab_QCD_Pt-300toInf_EMEnriched_TuneCUETP8M1_13TeV_pythia8_25ns/
CRAB project directory:     /nfs/home/fynu/obondu/Higgs/CMSSW_7_4_10/src/cp3_llbb/GridIn/test/tasks/crab_QCD_Pt-300toInf_EMEnriched_TuneCUETP8M1_13TeV_pythia8_25ns
Task name:          150923_134006:obondu_crab_QCD_Pt-300toInf_EMEnriched_TuneCUETP8M1_13TeV_pythia8_25ns
Task status:            FAILED  on schedd: crab3test-2@vocms095.cern.ch
Dashboard monitoring URL:   http://dashb-cms-job.cern.ch/dashboard/templates/task-analysis/#user=crab&refresh=0&table=Jobs&p=1&records=25&activemenu=2&status=&site=&tid=150923_134006%3Aobondu_crab_QCD_Pt-300toInf_EMEnriched_TuneCUETP8M1_13TeV_pythia8_25ns

Jobs status:            failed         20.0% ( 2/10)
                finished       60.0% ( 6/10)
                running        20.0% ( 2/10)

No publication information (publication has been disabled in the CRAB configuration file)

Error Summary: (use --verboseErrors for details about the errors)

 2 jobs failed with exit code 50664

Have a look at https://twiki.cern.ch/twiki/bin/viewauth/CMSPublic/JobExitCodes for a description of the exit codes.
Log file is /nfs/home/fynu/obondu/Higgs/CMSSW_7_4_10/src/cp3_llbb/GridIn/test/tasks/crab_QCD_Pt-300toInf_EMEnriched_TuneCUETP8M1_13TeV_pythia8_25ns/crab.log
```

Then running `runPostCrab.py`

```
>runPostCrab.py crab_QCD_Pt-300toInf_EMEnriched_TuneCUETP8M1_13TeV_pythia8_25ns.py
##### Get information out of the crab config file (work area, dataset, pset)
done

##### Check if the dataset exists in the database
calling das_import
****das query: [{u'dataset': [{u'name': u'/QCD_Pt-300toInf_EMEnriched_TuneCUETP8M1_13TeV_pythia8/RunIISpring15DR74-Asympt25ns_MCRUN2_74_V9-v2/MINIAODSIM', u'datatype': u'mc', u'creation_time': u'2015-06-18 12:13:25', u'nevents': 3696716, u'tag': None, u'size': 136056963110}]}]
Dataset #None:
  name: /QCD_Pt-300toInf_EMEnriched_TuneCUETP8M1_13TeV_pythia8/RunIISpring15DR74-Asympt25ns_MCRUN2_74_V9-v2/MINIAODSIM
  process: QCD_Pt-300toInf_EMEnriched_TuneCUETP8M1_13TeV_pythia8
  cross-section: 1.0
  number of events: 3696716
  size on disk: 136056963110
  CMSSW release: CMSSW_7_4_1_patch1
  global tag: MCRUN2_74_V9
  type (data or mc): mc
  center-of-mass energy: 13.0 TeV
  creation time (on DAS): 2015-06-18 12:13:25
  comment: 
Insert into the database? [y]|n: 
done
done

##### Get info from crab (outputs, report)
done

##### Get information from the output files
Warning: failed to retrieve number of entries for '/store/user/obondu/QCD_Pt-300toInf_EMEnriched_TuneCUETP8M1_13TeV_pythia8/QCD_Pt-300toInf_EMEnriched_TuneCUETP8M1_13TeV_pythia8_25ns/150923_134006/0000/output_mc_10.root'
∑w = 1929157.0000
Number of selected events: 19

##### Check if the job processed the whole sample
Warning: You are about to add in the DB a sample which has not been completely processed (1929157 events out of 3696716, 52.19%)
If you want to update this sample later on with more statistics, simply re-execute this script with the same arguments.

##### Figure out the code(s) version
FWUrl= https://github.com/cp3-llbb/Framework/tree/v0.1.1+74X
AnaUrl= https://github.com/cp3-llbb/HHAnalysis/tree/6fd46fc

##### Put it all together: write this sample into the database
Sample #None (created on None by obondu):
  name: QCD_Pt-300toInf_EMEnriched_TuneCUETP8M1_13TeV_pythia8_25ns_v0.1.1+74X_6fd46fc
  path: /storage/data/cms/store/user/obondu/QCD_Pt-300toInf_EMEnriched_TuneCUETP8M1_13TeV_pythia8/QCD_Pt-300toInf_EMEnriched_TuneCUETP8M1_13TeV_pythia8_25ns/150923_134006/0000
  type: NTUPLES
  number of processed events: 1929157
  number of events: 19
  normalization: 1.0
  sum of event weight: 1929157.0
  (effective) luminosity: None
  code version: https://github.com/cp3-llbb/HHAnalysis/tree/6fd46fc https://github.com/cp3-llbb/Framework/tree/v0.1.1+74X
  comment: Sample was not fully processed, only 1929157/3696716 events were processed
  source dataset: 2
  source sample: None
Insert into the database? [y]|n: 
```

now looking in the DB itself:

```
>search_SAMADhi.py sample
5   GluGluHToWWTo2L2Nu_M125_13TeV_powheg_JHUgen_pythia8_25ns_v0.1.1+74X-10-gf55f525_v0.1.1+74X-10-gf55f525
6   QCD_Pt-300toInf_EMEnriched_TuneCUETP8M1_13TeV_pythia8_25ns_v0.1.1+74X_6fd46fc
```

so that's almost a +1 from my side: incomplete status is handled correctly, warnings on empty files is good, database seemed filled, the only missing thing in my opinion is displaying the list of correct files when calling `search_SAMADhi.py`:

```
>search_SAMADhi.py sample -i 6 -l
Sample #6 (created on 2015-09-25 13:02:33 by obondu):
  name: QCD_Pt-300toInf_EMEnriched_TuneCUETP8M1_13TeV_pythia8_25ns_v0.1.1+74X_6fd46fc
  path: /storage/data/cms/store/user/obondu/QCD_Pt-300toInf_EMEnriched_TuneCUETP8M1_13TeV_pythia8/QCD_Pt-300toInf_EMEnriched_TuneCUETP8M1_13TeV_pythia8_25ns/150923_134006/0000
  type: NTUPLES
  number of processed events: 1929157
  number of events: 19
  normalization: 1.0
  sum of event weight: 1929160.0
  (effective) luminosity: 1929160.0
  code version: https://github.com/cp3-llbb/HHAnalysis/tree/6fd46fc https://github.com/cp3-llbb/Framework/tree/v0.1.1+74X
  comment: Sample was not fully processed, only 1929157/3696716 events were processed
  source dataset: 2
  source sample: None
--------------------------------------------------------------------------------------
```

the one possible other question I see with transitioning is: we would have to re-fill all the datasets in the database as well ?

**Comment by blinkseb** [comment]
Wahou thanks for the complete test :) It's good to see that it works as expected.

I'm currently working on displaying the list of files for the sample, however it's a bit more tricky than it sounds because of how storm works.

About the transition, that's a good question. It's fairly easy to update the current database to the new model, however, `event_weight_sum` and the files table will be empty. The ideal solution here is to re-execute the `runPostCrab.py` script on the crab tasks you want to update, don't you think?

I'll update this PR when I've made some progress on SAMADhi.

**Comment by OlivierBondu** [comment]
yep, sounds perfect !

**Comment by blinkseb** [comment]
Ok so I rebased this PR with improvements:
- Updating an existing dataset now works correctly (the list of files is correctly updated)
- The list of files is now also printed when using `search_SAMADhi.py`
- `crab getoutput --dump --xrootd` is not needed anymore.

If you want to test, you'll also need to update the SAMADhi repository.

Output:

```
##### Put it all together: write this sample into the database
A sample with the same name already exists in the database. Replace by:
Sample #9 (created on 2015-09-25 17:32:23 by sbrochet):
  name: GluGluHToWWTo2L2Nu_M125_13TeV_powheg_JHUgen_pythia8_25ns_v0.1.1+74X-10-gf55f525_v0.1.1+74X-10-gf55f525
  path: /storage/data/cms/store/user/sbrochet/GluGluHToWWTo2L2Nu_M125_13TeV_powheg_JHUgen_pythia8/GluGluHToWWTo2L2Nu_M125_13TeV_powheg_JHUgen_pythia8_25ns/150924_122611/0000
  type: NTUPLES
  number of processed events: 50000
  number of events: 30467
  normalization: 1.0
  sum of event weight: 50000.0
  (effective) luminosity: 50000.0
  code version:  
  comment: Sample was not fully processed, only 50000/499400 events were processed
  source dataset: 1
  source sample: None
  1 files: 
    - /store/user/sbrochet/GluGluHToWWTo2L2Nu_M125_13TeV_powheg_JHUgen_pythia8/GluGluHToWWTo2L2Nu_M125_13TeV_powheg_JHUgen_pythia8_25ns/150924_122611/0000/output_mc_1.root (30467 entries)

? [n]|y:
```

```
search_SAMADhi.py sample --long
Sample #9 (created on 2015-09-25 17:32:23 by sbrochet):
  name: GluGluHToWWTo2L2Nu_M125_13TeV_powheg_JHUgen_pythia8_25ns_v0.1.1+74X-10-gf55f525_v0.1.1+74X-10-gf55f525
  path: /storage/data/cms/store/user/sbrochet/GluGluHToWWTo2L2Nu_M125_13TeV_powheg_JHUgen_pythia8/GluGluHToWWTo2L2Nu_M125_13TeV_powheg_JHUgen_pythia8_25ns/150924_122611/0000
  type: NTUPLES
  number of processed events: 50000
  number of events: 30467
  normalization: 1.0
  sum of event weight: 50000.0
  (effective) luminosity: 50000.0
  code version:  
  comment: Sample was not fully processed, only 50000/499400 events were processed
  source dataset: 1
  source sample: None
  1 files: 
    - /store/user/sbrochet/GluGluHToWWTo2L2Nu_M125_13TeV_powheg_JHUgen_pythia8/GluGluHToWWTo2L2Nu_M125_13TeV_powheg_JHUgen_pythia8_25ns/150924_122611/0000/output_mc_1.root (30467 entries)
```

If it's fine with everybody, we'll need to upgrade the database once everything is merged. That'll need we should expect some downtime while the migration is happening, but should not take more than 5 minutes. I'll prepare a migration script.

**Comment by blinkseb** [comment]
I've created a PR in SAMADhi upstream: https://github.com/delaere/SAMADhi/pull/2

**Comment by OlivierBondu** [comment]
sounds good, since the PR in SAMADhi is merged, can we merge this one ? Or you prefer another round of testing ? (I can volunteer)

**Comment by blinkseb** [comment]
It should be fine to merge. Once merged, I'll update the current database layout to the new one, and, finger-crossed, everything should works :)

**Comment by blinkseb** [comment]
Database successfully upgraded!

**Comment by swertz** [comment]
Great :) What's to be done on the user side? Update SAMADhi and GridIn and that's it?

**Comment by blinkseb** [comment]
Exactly :) runPostCrab has been updated. You should be able to execute it
on old tasks to update the sample in the db.

2015-09-28 11:56 GMT+02:00 Sébastien Wertz notifications@github.com:

> Great :) What's to be done on the user side? Update SAMADhi and GridIn and
> that's it?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/cp3-llbb/GridIn/pull/27#issuecomment-143697680.

## 

Sébastien Brochet

**Pull request merged**